### PR TITLE
Add initial train script for fma genre classification

### PR DESCRIPTION
**Pull request by askerry**
A very basic first pass that gets us training on genre classification with tensorflow

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,9 +1,11 @@
-ipython==6.2.1
+ipython==5.2.1
```
Comment (rryan) [inline requirements.txt]: Should we drop version numbers entirely? This is due to differences between our environments.

File: requirements.txt [hunk 2]
```diff
@@ -1,9 +1,11 @@
...
 jupyter==1.0.0
 librosa==0.5.1.
 matplotlib==2.1.0
 numpy==1.13.3
 pandas==0.21.0
+pydot==1.2.3
```
Comment (rryan) [inline requirements.txt]: Want to separate these two out at the bottom and put a comment that it's for # fma_util.py ?

File: train.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
+"""Script for training a tensorflow model on audio datasets.
```
Comment (rryan) [inline train.py]: nit: TensorFlow

File: train.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+Usage:
+    python train.py --dataset=fma
```
Comment (rryan) [inline train.py]: Do you think we should put shebangs at hte top of files intended to be run and `chmod +x`, so we don't have to run python explicitly?

`#!/usr/bin/env python3` ?

File: train.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+"""
+
+import argparse
+import logging
+
+import tensorflow as tf
+
+from datasets import fma
+
+MINI_BATCH_SIZE = 32
```
Comment (rryan) [inline train.py]: Whiel you're at it, should we make these flags instead of constants?

File: train.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+LEARNING_RATE = .5
+
+
+def load_data(dataset, split="training", size="small"):
```
Comment (rryan) [inline train.py]: split/size as flags?

File: train.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    """Load data into trainable iterator."""
+    logging.info("Loading dataset %s" % dataset)
+    # TODO(amy): build support for cross validation rather than
+    # static train/test split
+    if dataset == 'fma':
+        path_to_data = 'raw_data'
```
Comment (rryan) [inline train.py]: maybe make this a flag?

File: train.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        dataset, all_labels = fma.read(path_to_data, split, size)
+        # Shuffle with shuffle buffer 10x the batch size
```
Comment (rryan) [inline train.py]: nit: end in period, here and below?

File: train.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        dataset = dataset.shuffle(MINI_BATCH_SIZE * 10)
+        batched_dataset = dataset.batch(MINI_BATCH_SIZE)
+        # Create iterator that will make one pass through the dataset
+        iterator = batched_dataset.make_one_shot_iterator()
+        # Returns placeholder tensors (features and labels) representing
```
Comment (rryan) [inline train.py]: placeholder has a specific meaning in TF (tf.placeholder), so I might phrase this differently. What about:

Returns features and labels dictionaries containing tensors from this mini-batch. See `fma.read` for details on the tensors in these dictionaries.

File: train.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        labels = y['genre_top']
+    else:
+        raise ValueError('Unknown dataset %s' % dataset)
+    return features, labels, all_labels
+
+
+# TODO(amy): figure out how we want to handle flexibly specifying the
+# network structure
+def predict(features, num_labels):
+    """Construct actual network to predict labels from features"""
```
Comment (rryan) [inline train.py]: In TF shapes are really hard to keep in your head, so I find it helpful to add full-blown docstrings.
How about: 
```
Args:
  features: A [batch_size, input_feature_dimension] tf.float32 tensor of features to predict from.
  num_labels: A Python integer indicating the total number of labels.

Returns:
  A [batch_size, num_labels] tf.float32 of unnormalized predictions.
```

File: train.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    # create a single layer network returning unnormalized probability
+    # of each class
+    logits = tf.contrib.layers.fully_connected(
+        features, num_labels, activation_fn=tf.nn.relu)
+    return logits
+
+
+def specify_model(features, labels, all_labels, num_oov_buckets=1):
+    """Specify the loss and structure of the model."""
```
Comment (rryan) [inline train.py]: Same comment RE: docstring. I think it helps readability a lot to have a full-blown one, sicne it's easy to forget what things are -- e.g. that `labels` is a string tensor or already int, etc. 

```
Args:
  features: A [batch_size, input_feature_dimension] tf.float32 tensor of features to predict from.
  labels: A [batch_size] tf.string tensor of labels.
  all_labels: A Python list of strings containing all the labels in the dataset.
  num_oov_buckets: The number of out-of-vocabulary buckets to create for handling labels that are not in all_labels.

Returns:
  train_step: A tf.Operation that minimizes the loss.
  loss: A scalar tf.float32 tensor that is the total loss for the current mini-batch.
```

File: train.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        LEARNING_RATE).minimize(cross_entropy)
+    return train_step, cross_entropy
+
+
+def train_loop(train_op, loss, num_iterations=1000):
+    """Core training loop."""
+    # TODO(amy): use MonitoredTrainingSession instead of raw session
+    try:
+        with tf.Session() as session:
+            # initialize all variables in the graph
```
Comment (rryan) [inline train.py]: nit: capitals/periods.

File: train.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    y_one_hot = tf.one_hot(y_ids, num_labels)
+    logits = predict(features, num_labels)
+    cross_entropy = tf.losses.softmax_cross_entropy(y_one_hot, logits)
+    # compute the gradients and creates a tf op minimize the loss
+    train_step = tf.train.GradientDescentOptimizer(
+        LEARNING_RATE).minimize(cross_entropy)
+    return train_step, cross_entropy
+
+
+def train_loop(train_op, loss, num_iterations=1000):
```
Comment (rryan) [inline train.py]: should num_iterations be a flag?

**Review by rryan — APPROVED** [review]
It lives! 

LGTM, but I suggested a bunch of cleanups. Feel free to push back if you want to move forward.

**Review by rryan — APPROVED** [review]
